gdal 2.0.0

mkrapp opened the pull request.

mikemcquaid commented on Library/Formula/gdal.rb:
```diff
@@ -1,8 +1,8 @@
-  sha256 "66bc8192d24e314a66ed69285186d46e6999beb44fc97eeb9c76d82a117c0845"
+  url "http://download.osgeo.org/gdal/2.0.0/gdal-2.0.0.tar.gz"
+  sha256 "53761563ff53c5bf27bff7c4d6cab8bb1634baccefda05348e0f3b7acaf4c9e6"
   revision 2
```
This `revision` will need to be removed as it's a version update, sorry!

mkrapp replied:
Remove the whole line?

mikemcquaid replied:
Yup, thanks!

mkrapp commented:
I've updated to 2.0.0 because I also had the same issue #42031.

mkrapp commented:
From the tests, I see that `brew test --verbose points2grid` fails because the `gdal` library cannot be found `/usr/local/lib/libgdal.1.dylib`. The link to the library now being `/usr/local/lib/libgdal.20.dylib`.

Setting `ln -s /usr/local/lib/libgdal.dylib /usr/local/lib/libgdal.1.dylib` and the `points2grid` test passes.

Any idea how to fix that?

mikemcquaid commented:
@mkrapp everything output by `brew uses gdal` probably needs a `revision 1` line added (or that number incremented if the line is already there.

DomT4 commented:
Yeah, when we looked at this before it was a revision party situation. If there's a ton of `gdal` stuff in other Homebrew core taps it's probably courteous to give them a heads up on this as well.

mkrapp commented:
@mikemcquaid should I check and go for it? Or don't we upgrade to 2.0.0?

DomT4 commented:
Add or update the revision lines on each of these formulae. Use a unique commit (In the same PR) for each updated formula with the commit message along the lines of `formula_name: revision for gdal`, Thanks!

```
gdal-grass
grass
liblas
mapnik
mapserver
open-scene-graph
orfeo
pdal
points2grid
postgis
simple-tiles
```

mkrapp commented:
I start with the 3 formulas from the tests (`pdal`, `points2grid`, and `simple-tiles`) to see if that works then, `simple-tiles` still fails with 

    ERROR 1: Stub implementation in GDAL 2.0
    ERROR 1: Stub implementation in GDAL 2.0

dunn commented:
Tap formulae that may need revisioning:
```
homebrew/science/gmt
homebrew/science/gmt4
homebrew/science/osgearth
homebrew/science/qgis
homebrew/science/sumo
homebrew/versions/postgis20
homebrew/x11/xastir
```

mkrapp commented:
@dunn Could we create another pull request for taps and just update formulae in `homebrew`?

bfontaine commented:
@mkrapp You’ll have to make one PR per affected tap.

mkrapp commented:
@bfontaine I'm so looking forward to this. BTW, how can I test if the new version of `gdal` does not break any other formula? And the revision flag just indicates that some of the dependencies have been updated, right?

bfontaine commented:
> how can I test if the new version of gdal does not break any other formula?

Install the new version of  `gdal` locally, then install formulae that depend on it.

> And the `revision` flag (or increment) just indicates that some of the dependencies have been updated, right?

It forces a reinstall of the formula, so it’ll be built against the new `gdal` version.

mkrapp commented:
I tried to install the dependent formulae, some failed:

`gdal-grass` passes with patch
`grass` passes
`liblas` passes
`mapnik` passes
`mapserver` fails
`open-scene-graph` OK with update to `3.4.0`; fails for `3.3.3`
`orfeo` fails 
`pdal` passes
`points2grid` passes
`postgis` passes
`simple-tiles` passes; test still fails with `ERROR 1: Stub implementation in GDAL 2.0` (see above)

There is a lot going wrong, are you sure all those packages worked for `gdal` < `2.0.0`? Just asking...

What's you're advise on how to proceed?

mkrapp commented:
Here's a quick update on my latest commits:

* updated `orfeo` to `5.0.0` (needs `ossim`, which I copied, modified from [here](https://github.com/imintel/homebrew-imintel/blob/master/ossim.rb) and added to my repo)
* updated `mapserver` to `7.0.0` (builds with `cmake`)

All packages pass compilation on my machine.

mkrapp commented:
Is there a way to increase the timeout of the build bot?

`Build timed out (after 65 minutes)`

mikemcquaid commented:
We can but fixing `postgis` and `ossim` is probably worth doing before we bump the timeout.

DomT4 commented:
If we could get comments on any patches used as well, with upstream bug submissions/PRs/commits etc that'd be great. One commit per changed formula would be a lot easier to keep on top of, if at all possible.

mkrapp commented:
@mikemcquaid agreed!

 Installing `postgis` on my machine (yosemite) works well, both

`brew install --verbose --build-bottle postgis`

and 

`brew install postgis --verbose`.

The only error related to `postgis` I find on Jenkins' console output is 

```
syslog -F '$((Time)(local)) $(Sender)[$(PID)]: $Message' -k Time ge 1440081439 -k Sender kernel -o -k Time ge 1440081439 -k Sender sandboxd | grep deny > /Users/brew/Jenkins/prs/version/yosemite/logs/postgis/sandbox.build.log 
Error: Operation not permitted - (stage/usr/local/Cellar/postgresql/9.4.4/lib/postgis-2.1.so, /usr/local/Cellar/postgresql/9.4.4/lib/postgis-2.1.so)
```

Now checking `ossim`, seems like a missing `GEOS` installation.

mkrapp commented:
@DomT4 I had to patch `ogrgrass.h` for `gdal-grass` because of a wrong type that gave an error. Probably needs a fix in the `gdal-grass` source (for which I haven't submitted a bug report, yet):

`GetFeatureCount( int )` in `ogrgrass.h` returns `int` but should return `GIntBig` instead. Error message when compiling the source (`gdal-grass-1.4.3.tar.gz`) is:

```
In file included from ogrgrassdriver.cpp:30:
./ogrgrass.h:54:25: error: virtual function 'GetFeatureCount' has a different return type ('int') than the function it overrides
      (which has return type 'GIntBig' (aka 'long long'))
    int                 GetFeatureCount( int );
    ~~~                 ^

```

Regarding the commits, I'm pretty sure I commit one change at a time.

mkrapp commented:
That build failed and I don't know why? What does `ref HEAD is not a symbolic ref` mean?
```
Error Message

failed: brew pull --clean https://github.com/Homebrew/homebrew/pull/42845
Stacktrace

        fatal: ref HEAD is not a symbolic ref
==> Applying patch
Applying: update to gdal 2.0.0
Using index info to reconstruct a base tree...
M	Library/Formula/gdal.rb
Falling back to patching base and 3-way merge...
Auto-merging Library/Formula/gdal.rb
CONFLICT (content): Merge conflict in Library/Formula/gdal.rb
Failed to merge in the changes.
Patch failed at 0001 update to gdal 2.0.0
The copy of the patch that failed is found in:
   /usr/local/.git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
Error: Patch failed to apply: aborted.
```

mikemcquaid commented:
It looks like there may be a merge conflict.

mikemcquaid commented:
But I've retried the build to make sure.

DomT4 commented:
>  (for which I haven't submitted a bug report, yet):

Please do so and add that bug report above the patch in formula, Thanks!

> It looks like there may be a merge conflict.

Apparently so. This needs to be rebased on an up-to-date master branch before our CI can look over it again.

mkrapp commented:
@DomT4 done!

mkrapp commented:
BTW with `orfeo` updated to 5.0.0 this PR should be referenced: #41363

mkrapp commented:
If somebody can get back to this, I'm still confused what to do about the merge conflict and how to solve it (for CI).

bfontaine commented:
@mkrapp I guess you can cancel your merge commit in e32812d (`git reset --hard HEAD^`) then rebase on top of the master branch (`git rebase master`) and force-push again (`git push -f`)

mkrapp commented:
Am I too stupid? Probably, but anyhow. Any advise is appreciated to circumvent this "merge" issue.

mikemcquaid commented:
Squash commits by running `brew update; git checkout $YOUR_BRANCH; git rebase --interactive origin/master` and then mark each commit other than the first as `squash` or `fixup` then `git push -f`. Thanks!

elgertam commented:
I saw this was open and was wondering the status. Is this something I can help with?

mkrapp commented:
indeed, some help would be appreciated @elgertam. I'm struggling to understand the multitude iof (error/FAILED) build messages.

elgertam commented:
Understood, @mkrapp. I'll start by attempting to build on my machine. This is my first attempt contributing to Homebrew, so bear with me.

MatzFan commented:
Happy to help with testing & get the gdal 2.0 formula out. If this may be useful please let me have build/test instructions. Am currently running OS X 10.9.5

dunn commented:
Oop, now this'll need to be rebased.

mkrapp commented:
Rebase done! Now, let's see...

dunn commented on Library/Formula/gdal.rb:
```diff
@@ -79,11 +79,14 @@ class Gdal < Formula
     depends_on "ant" => :build
     depends_on "swig" => :build
   end
+<<<<<<< HEAD
```
You forgot to resolve a conflict here :see_no_evil:

dunn commented:
This will probably hit the timeout again, so wait to push a fix until we can schedule a build.

mkrapp commented:
@dunn. Thanks I fixed that and rebased. How do I know when to push a commit?

dunn commented:
@mkrapp The queue is currently empty so if you're online, push now.

dunn commented:
We've got test results from 4 hours ago, actually: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/34175/#showFailuresLink

DomT4 commented:
On `postgis` you'll need to see/follow the notes in #41962 to get that to pass.

mkrapp commented:
Many installs seem to fail because of `open-scene-graph`. I couldn't build it on El Capitan.

elgertam commented:
@mkrapp Which branch are you working out of, mkrapp or master?

mkrapp commented:
Hi everybody,

this PR has grown more complex over time and I was wondering if we could break it down somehow to resolve the issues with failed CI's. Here are some comments:

* Is it possible, for example, to test the updates in a kind of "sandbox" environment?
* Can some packages other than `gdal` be updated separately? For example, update `map-server` to `7.0.0` with `gdal 1.11.3`, or add `ossim`, then update `orfeo` and so forth. Same for some of the other packages.

elgertam commented:
@mkrapp +1

I tested your changes last night, and I'm inclined to agree. I tested last night, and GDAL built and installed just fine. I had some problems getting the gdal-grass formula to install because grass was failing to install, which apparently isn't uncommon.

DomT4 commented:
`ossim` can become its own PR after this is merged. We should try and get `postgis` fixed outside this PR before moving forwards on this too much further as the same set of errors related to that are causing a lot of the failures. The OSG failure is possibly related to #44524.

Sorting out B & C on the above list before we push ahead on this should fix almost all the errors you're seeing, theoretically.

bfontaine commented:
Ping?

elgertam commented:
@bfontaine and @DomT4: What else is there to be done here in order to accept the PR? I'm able to build GDAL using @mkrapp's PR, and the only errors I have seen are unrelated to GDAL itself.

bfontaine commented:
@elgertam The errors you see are formulae that depend on GDAL; we need to fix all of them before merging this PR.

mkrapp commented:
Any idea how to move forward from here?

bfontaine commented:
This branch needs to be rebased on top of `master`.

mkrapp commented:
Hi. Here is a summary of the most recent build failure(s) and what I did to try fixing it. I only checked `10.11.1` (El Capitan)

* `pdal` did not build:
```
dyld: Library not loaded: /usr/local/opt/gdal/lib/libgdal.1.dylib
  Referenced from: /usr/local/Cellar/pdal/0.9.9_1/bin/pdal
  Reason: image not found
Error: pdal: failed
Failed executing: /usr/local/Cellar/pdal/0.9.9_1/bin/pdal info
```
* `points2grid` did not build:
```
dyld: Library not loaded: /usr/local/opt/gdal/lib/libgdal.1.dylib
  Referenced from: /usr/local/Cellar/points2grid/1.3.0_2/bin/points2grid
  Reason: image not found
Error: points2grid: failed
Failed executing: /usr/local/Cellar/points2grid/1.3.0_2/bin/points2grid -i /usr/local/Cellar/points2grid/1.3.0_2/libexec/example.las -o example --max --output_format grid
```
:bulb: However, I'm able to fix the above two errors manually by creating a symbolic link in `/usr/local/opt/gdal/lib`
```
ln -s libgdal.dylib libgdal.1.dylib
```
* `open-scene-graph` did not build:
```
Error: open-scene-graph 3.4.0 did not build
```
:bulb: If I install it manually everything works fine with `cmake` options taken from output of `brew install open-scene-graph --with-gdal --verbose`:
```
wget http://trac.openscenegraph.org/downloads/developer_releases/OpenSceneGraph-3.4.0.zip
unzip OpenSceneGraph-3.4.0.zip
cd OpenSceneGraph-3.4.0
mkdir build && cd build
cmake .. -DCMAKE_C_FLAGS_RELEASE= -DCMAKE_CXX_FLAGS_RELEASE= -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/open-scene-graph/3.4.0 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DBUILD_DOCUMENTATION=OFF -DCMAKE_OSX_ARCHITECTURES=x86_64 -DOSG_DEFAULT_IMAGE_PLUGIN_FOR_OSX=imageio -DOSG_WINDOWING_SYSTEM=Cocoa
make && make install
```
A while ago I also commented on a referenced issue #44524.
* `grass` did not build:
Seems like `wxpython` causes some issues
```
Error: wxpython 3.0.2.0 did not build
```
Issues with `wxpython` are #45381 and #45024. :bulb: However, I was able to install it locally.
* `gdal-grass` did not build because of `wxpython`
* `ossim` did not build because of `open-scene-graph`
* `orfeo` did not build because of `open-scene-graph`

So we can break it down to `open-scene-graph` and `wxpython` (and some missing links of the `gdal` library).
:question: Any ideas how to resolve that?

mikemcquaid commented:
> pdal did not build:
> points2grid did not build:

These two need a `revision` bump rather than symlink. Not sure about the others.

mkrapp commented:
OK, `pdal` and `points2grid` are no longer an issue.

But `ossim` caused trouble, which I fixed with an updated version (`1.8.20-1`).

Also, while checking errors in `open-scene-graph` I decided to switch to the latest version (`3.5.1`). I cleaned up the `open-scene-graph` formula but still some errors persist. One is related to a c++11-narrowing error, see #46372, the other is an error in `open-scene-graph` itself: 
```
OpenSceneGraph-3.5.1/include/osgViewer/View:130:91: error: member reference base type
      'const osg::ref_ptr<T> *' is not a structure or union
        template<class T> void setImagePager(const osg::ref_ptr<T>* ip) { setImagePager(ip.get()); }
                                                                                        ~~^~~~
1 error generated.
```
A patched the corresponding file and sent a bug report to the osg-users mailing list.

That formula now installs on my machine (`10.11.1`). Crossing fingers for the build...

mkrapp commented:
Finally, `open-scene-graph` was built. :joy:

Firefishy commented:
Latest is gdal 2.0.1: http://download.osgeo.org/gdal/2.0.1/gdal-2.0.1.tar.gz
b55f794768e104a2fd0304eaa61bb8bda3dc7c4e14f2c9d0913baca3e55b83ab

DomT4 commented on Library/Formula/mapserver.rb:
```diff
@@ -29,104 +29,29 @@ class Mapserver < Formula
-      "--with-wfs"
+    args = std_cmake_args
+    args += [
+      "-DCMAKE_BUILD_TYPE=Release",
```
We pass this already as part of `std_cmake_args`.

DomT4 commented on Library/Formula/mapserver.rb:
```diff
@@ -29,104 +29,29 @@ class Mapserver < Formula
-
   test do
-    system "#{bin}/mapserver-config", "--version"
+    system "mapserv", "-v"
```
Please don't remove the `#{bin}` here.

DomT4 commented on Library/Formula/open-scene-graph.rb:
```diff
@@ -33,8 +33,6 @@ class OpenSceneGraph < Formula
 
-  # patch necessary to ensure support for gtkglext-quartz
-  # filed as an issue to the developers https://github.com/openscenegraph/osg/issues/34
   patch :DATA
```
Can you add a link/explanation as to what the new patch is, where it's from, etc? Thanks.

DomT4 commented on Library/Formula/orfeo.rb:
```diff
@@ -1,35 +1,25 @@
+  url "https://downloads.sourceforge.net/project/orfeo-toolbox/OTB/OTB-5.0.0/OTB-5.0.0.tgz"
+  sha256 "35e2f1d280f4a33b3d38af840a83069b34b2ea99acba18b1a944acccdbf1a976"
+
+  option "with-examples", "Compile and install various examples"
```
You'll need to deprecate the old options onto the new options here with the:

```ruby
deprecated_option "examples" => "with-examples"
```
style syntax.

DomT4 commented on Library/Formula/orfeo.rb:
```diff
@@ -1,35 +1,25 @@
-  option "java", "Enable Java support"
-  option "patented", "Enable patented algorithms"
+  depends_on "open-scene-graph" => ["with-gdal"]
+  depends_on "homebrew/science/insighttoolkit"
```
Can't have mandatory cross-tap dependencies in the core I'm afraid.

DomT4 commented on Library/Formula/orfeo.rb:
```diff
@@ -1,35 +1,25 @@
-
     mkdir "build" do
-      system "cmake", "..", *args
+      system "cmake", ".."
```
Presumably still want `std_cmake_args` here?

DomT4 commented on Library/Formula/ossim.rb:
```diff
@@ -0,0 +1,71 @@
+  def install
+    args = std_cmake_args
+    args += [
+      "-DCMAKE_BUILD_TYPE=Release",
```
We pass this and the install prefix as part of `std_cmake_args`.

DomT4 commented on Library/Formula/ossim.rb:
```diff
@@ -0,0 +1,71 @@
+    args << "-DBUILD_OSSIMGDAL_PLUGIN=" + (build.with?("gdal") ? "ON" : "OFF")
+    args << "-DBUILD_OSSIMPNG_PLUGIN=" + (build.with?("libpng") ? "ON" : "OFF")
+
+    ossim_dev_home = Dir.pwd
```
Would `Dir.pwd` not be `buildpath` here?

DomT4 commented on Library/Formula/ossim.rb:
```diff
@@ -0,0 +1,71 @@
+      system "make", "install"
+    end
+
+    Dir.mkdir("#{prefix}/etc")
```
Don't need to throw out to `Dir` here. Just `(prefix/"etc").mkpath`

DomT4 commented on Library/Formula/ossim.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  test do
+    system "ossim-info", "--version"
```
Please use the `#{bin}` style links to the executables here.

mkrapp commented:
@DomT4 Regarding the `orfeo` dependencies. It seems like ITK aka `homebrew/sience/insighttoolkit` is [mandatory](https://www.orfeo-toolbox.org/SoftwareGuide/SoftwareGuidech2.html#x14-220071):

```
CMake Error at build/lib/cmake/OTB-5.0/Modules/OTBITK.cmake:6 (find_package):
  Could not find a package configuration file provided by "ITK" with any of
  the following names:

    ITKConfig.cmake
    itk-config.cmake

  Add the installation prefix of "ITK" to CMAKE_PREFIX_PATH or set "ITK_DIR"
  to a directory containing one of the above files.  If "ITK" provides a
  separate development package or SDK, be sure it has been installed.
```
Also, `homebrew/science/insighttoolkit` needs `homebrew/science/vtk`.

Is it possible to move `orfeo` from the core to `homebrew/science`?

dmwelch commented:
FYI: `orfeo` is built on top of `ITK` - `ITK` and `VTK` were written by a medical image processing consortium and `orfeo` takes many of their objects as base classes. 🤓

nextstopsun commented:
Is this ever going to happen? So many people are waiting for gdal 2 to be installed via homebrew, it's really a fundamental spatial library. Sadly, osgeo4mac doesn't have any updates on gdal either. Please don't give up guys, it's definitely worth giving another try.

elgertam commented:
I concur with @nextstopsun. It appears that, other than a few style and convention issues, we are hung up on the `orfeo` package, which may depend on packages in the `homebrew/science` tap. Can @DomT4 or another homebrew maintainer help us figure out the solution to this: either moving `orfeo` itself to `homebrew/science` or figuring out a way to depend solely on the core packages?

I'm willing to help in any way I can.

@mkrapp: apart from the question about `orfeo`, what else must be done?

elgertam commented:
For what it's worth, this is what I get when trying to install orfeo with vanilla homebrew:

```
$ brew install orfeo
fltk: This formula either does not compile or function as expected on OS X
versions newer than Yosemite due to an upstream incompatibility.
Error: An unsatisfied requirement failed this build.```

Seems to me that moving it to `homebrew/science` would serve everyone best.

mikemcquaid commented:
> Seems to me that moving it to homebrew/science would serve everyone best.

:+1:

nextstopsun commented:
Although orfeo has some really neat image classification functions, dumping it to a science tap seems like a sacrifice for greater good. But what is wrong with fltk?

DomT4 commented:
> But what is wrong with fltk?

As it says on the tin:

> fltk: This formula either does not compile or function as expected on OS X
versions newer than Yosemite due to an upstream incompatibility.

It's already fixed upstream. We're waiting on a new stable release with the fixes included.

bfontaine commented:
Anything new here?

nextstopsun commented:
Since new postgresql 9.5 and postgis 2.2.1 are released, gdal 2 is on high demand. Still waiting for it very very much.

mkrapp commented:
Can somebody take the lead on moving `orfeo` to `home-brew/science`?

nextstopsun commented:
Any news on this issue?

nijikon commented:
What is pending here?

nextstopsun commented:
Still no progress on this issue?

mikemcquaid commented:
@nijikon The build is failing because someone needs to move `orfeo` to `homebrew/science`, there's a merge conflict and possibly other things.

@nextstopsun No. We'd have posted on this issue if there was. If you would like to create your own pull requests to address the issues above: that'd be great, thanks.

oeon commented:
There's another pull request for gdal 2+ over at the homebrew-osgeo4mac tap https://github.com/OSGeo/homebrew-osgeo4mac/pull/122 - I thought I would make both camps aware.
